# HTAN Bulk RNA Expression pipeline
The pipeline aligns bulk RNA-seq reads and produces aligned BAMs, readcount, and FPKM.



## Release



## Output format
Each sample gets its TSV in the exact same gene order. The output TSV file has the following columns:

|   Column   |             Description              |
| :--------- | :----------------------------------- |
| gene_id    | Ensembl gene ID                      |
| symbol     | Gene symbol (not unique)             |
| read_count | stranded read count by featureCounts |
| fpkm       | FPKM                                 |
| fpkm_uq    | FPKM-UQ                              |

Each batch of execution will produce a TSV `analysis_summary.dat` containing all output with the following columns:



## Processing description

### Alignment
STAR v2.7.4a and GDC's genome reference [GRCh38.d1.vd1][GDC Reference Files].

### Readcount
The readcount is generated by featureCounts (subread v2.0.1) under stranded mode with parameters: `-g gene_id -t exon -Q 10 -p -B`. The readcount is later converted to FPKM and FPKM-UQ using [GDC's formula].

### Annotation
[GENCODE v34 GTF][gencode-gtf]. Refer to the `prepare_annotation/` folder for how the annotation files were generated.

[GDC Reference Files]: https://gdc.cancer.gov/about-data/data-harmonization-and-generation/gdc-reference-files
[gencode-gtf]: ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/gencode.v34.annotation.gtf.gz
[GDC's formula]: https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/Expression_mRNA_Pipeline/#upper-quartile-fpkm



## Setup
```
conda create -n htan_bulk_rna python=3.8 \
    snakemake-minimal=5.19.1 \
    pandas=1.0.4 \
    star=2.7.4a \
    samtools=1.10 htslib=1.10 \
    subread=2.0.1
```


